### PR TITLE
3.0.0-Modal

### PR DESCRIPTION
**Pull request by CloudPower97**
# Description

This Pull Request is for #599 .

Dropping __jQuery__ support.

Adding new options (as per the [documentation](https://materializecss.com/modals.html#options)):

- onOpenStart
- onOpenEnd
- onCloseStart
- onCloseEnd
- preventScrolling

Removed deprecated options:

-  ready
- complete

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Updated spec to use the new mocker.

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have not generated a new package version. (the maintainers will handle that)

### CODE REVIEW

**Comment by CloudPower97** [comment]
@alextrastero can you take a look at `controlled modal with `open` prop` test? It's driving me crazy because it says that `this.instance.open()` is not a function.

But it can't be right, as matter of fact it does indeed work when used in production AND `initializes with options` test simulate a click on a button which in turns call the `hide` method....

Surprise surprise it doesn't complain when about `this.instance.close()`.

**Comment by alextrastero** [comment]
Will check ASAP

File: test/Modal.spec.js [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
+import mocker from './helper/new-mocker';
 
 describe('<Modal />', () => {
   let wrapper;
-  const modalMock = jest.fn();
-  const restore = mocker('modal', modalMock);
+  const modalInitMock = jest.fn();
+  const modalInstanceDestroyMock = jest.fn();
+  const modalMock = {
+    init: (el, options) => {
```
Comment (alextrastero) [inline test/Modal.spec.js]: you would need to add here `open` method mock
```js
const modalOpenMock = jest.fn();
const modalMock = {
  open: modalOpenMock
  ...
```

Comment (CloudPower97) [reply test/Modal.spec.js]: Hey @alextrastero, I've done that, but It still throws me the same error. Please, check the last commit

File: test/Modal.spec.js [hunk 2]
```diff
@@ -2,6 +2,9 @@ import React from 'react';
 import { shallow, mount } from 'enzyme';
 import Modal from '../src/Modal';
 import mocker from './helper/new-mocker';
+import idgen from '../src/idgen';
+
+jest.mock('../src/idgen');
```
Comment (alextrastero) [inline test/Modal.spec.js]: i've mocked the `idgen` so that all modals have the same ID

File: test/Modal.spec.js [hunk 3]
```diff
@@ -13,16 +16,18 @@ describe('<Modal />', () => {
     init: (el, options) => {
       modalInitMock(options);
       return {
+        open: modalOpenMock,
```
Comment (alextrastero) [inline test/Modal.spec.js]: open and close are the methods returned by `init` :)

File: test/Modal.spec.js [hunk 4]
```diff
@@ -65,7 +69,6 @@ describe('<Modal />', () => {
 
     afterEach(() => {
       modalInitMock.mockClear();
-      document.body.removeChild(document.body.lastElementChild);
```
Comment (alextrastero) [inline test/Modal.spec.js]: this shouldn't be needed since the component should auto clean itself

File: test/__snapshots__/Modal.spec.js.snap [hunk 5]
```diff
@@ -1,5 +1,11 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`<Modal /> controlled modal with \`open\` prop mounts opened 1`] = `
+<div>
+  <Portal />
+</div>
```
Comment (alextrastero) [inline test/__snapshots__/Modal.spec.js.snap]: this snapshot is updated since it's now shallowed

File: test/Modal.spec.js (hunk 3 above)
Comment (CloudPower97) [reply test/Modal.spec.js]: Ah, I see... Too bad I didn't catch this one! :grimacing:

**Pull request merged**